module/zfs/arc.c

```diff
@@ -8700,6 +9168,838 @@ l2arc_stop(void)
 	mutex_exit(&l2arc_feed_thr_lock);
 }
 
+/*
+ * Punches out rebuild threads for the L2ARC devices in a spa. This should
+ * be called after pool import from the spa async thread, since starting
+ * these threads directly from spa_import() will make them part of the
+ * "zpool import" context and delay process exit (and thus pool import).
+ */
+void
+l2arc_spa_rebuild_start(spa_t *spa)
+{
+	ASSERT(MUTEX_HELD(&spa_namespace_lock));
+
+	/*
+	 * Locate the spa's l2arc devices and kick off rebuild threads.
+	 */
+	for (int i = 0; i < spa->spa_l2cache.sav_count; i++) {
+		l2arc_dev_t *dev =
+		    l2arc_vdev_get(spa->spa_l2cache.sav_vdevs[i]);
+		if (dev == NULL) {
+			/* Don't attempt a rebuild if the vdev is UNAVAIL */
+			continue;
+		}
+		if (dev->l2ad_rebuild && !dev->l2ad_rebuild_cancel) {
+			(void) thread_create(NULL, 0,
+			    (void (*)(void *))l2arc_dev_rebuild_start,
+			    dev, 0, &p0, TS_RUN, minclsyspri);
+		}
+	}
+}
+
+/*
+ * Main entry point for L2ARC rebuilding.
+ */
+static void
+l2arc_dev_rebuild_start(l2arc_dev_t *dev)
+{
+	if (!dev->l2ad_rebuild_cancel) {
+		VERIFY(dev->l2ad_rebuild);
+		dev->l2ad_rebuild_began = B_TRUE;
+		(void) l2arc_rebuild(dev);
+		dev->l2ad_rebuild_began = B_FALSE;
+		dev->l2ad_rebuild = B_FALSE;
```

Comment:
I really appreciate having another set of eyes on these mechanisms.

`l2arc_remove_vdev()` checks for `l2ad_rebuild_began` at the very beginning. If true it sets `l2arc_rebuild_cancel` to true and it waits for the rebuild to be cancelled in `l2arc_rebuild()`, lines 9336-9342 in `module/zfs/arc.c`. This happens before the device is actually removed, see lines 9068-9077 in `module/zfs/arc.c`.

Do you think this is enough?

Reply:
I think I see what you mean. If `l2arc_remove_vdev()` checks for `l2ad_rebuild_began()` but it is still false, then it will proceed to remove the device. However nothing prevents `l2arc_rebuild()` from being called in that case. Will look into it.